File: freemazon/src/com/googlecode/freemazon/BookDownloadData.java
package com.googlecode.freemazon;

import java.io.Serializable;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Set;

/**
 * Persistent data regarding a single book download process.
 * @author DL
 */
public class BookDownloadData implements Serializable {
    public BookDownloadData(String bookTitle, int numPages) {
        this.bookTitle = bookTitle;
        for (int i = 1; i <= numPages; ++i) {
            unseenPages.add(i);
        }
    }

    public String bookTitle() {
        return bookTitle;
    }

    public boolean usedQuery(String query) {
        return usedQueries.contains(query);
    }

    public void onQuery(String query) {
        usedQueries.add(query);
    }

    public void onPageTokens(Map<Integer, String> newPageTokens) {
        for (Entry<Integer, String> entry : newPageTokens.entrySet()) {
            onPageToken(entry.getKey(), entry.getValue());
        }
    }

    public void onPageToken(Integer page, String token) {
        if ((!gotPages.contains(page)) && (!pageImages.containsKey(page))) {
            unseenPages.remove(page);
            pageTokens.put(page, token);
        }
    }

    public void onPageImages(Map<Integer, String> newPageImages) {
        for (Entry<Integer, String> entry : newPageImages.entrySet()) {
            final Integer page = entry.getKey();
            if (!gotPages.contains(page)) {
                unseenPages.remove(page);
                pageTokens.remove(page);
                pageImages.put(page, entry.getValue());
            }
        }
    }

    public void onGotPage(Integer page) {
        if (!gotPages.contains(page)) {
            unseenPages.remove(page);
            pageTokens.remove(page);
            pageImages.remove(page);
            gotPages.add(page);
        }
    }

    public Map<Integer, String> pageTokens() {
        return new HashMap<Integer, String>(pageTokens);
    }

    public Map<Integer, String> pageImages() {
        return new HashMap<Integer, String>(pageImages);
    }

    @Override
    public String toString() {
        return "unseen: " + unseenPages.size() +
                ", tokens: " + pageTokens.size() +
                ", img urls: " + pageImages.size() +
                ", got: " + gotPages.size();
    }
    
    /** Pages with unknown image url or token. */
    private Set<Integer> unseenPages = new HashSet<Integer>();

    /** Pages with known tokens. */
    private Map<Integer, String> pageTokens = new HashMap<Integer, String>();

    /** Pages with known image url. */
    private Map<Integer, String> pageImages = new HashMap<Integer, String>();

    /** Downloaded pages. */
    private Set<Integer> gotPages = new HashSet<Integer>();

    /** Search queries used. */
    private Set<String> usedQueries = new HashSet<String>();

    /** Book name. */
    private String bookTitle = "book";
    
    private static final long serialVersionUID = 1L;
}

File: src/BookDownloadData.js
import _ from 'lodash';

export default class BookDownloadData {
	constructor({bookTitle = '', numPages = 0}) {
		this._bookTitle = bookTitle;
		
		this._usedQueries = {};
		this._suggestedQueries = {};
		
		this._unseenPages = {};
		for (let i = 1; i < numPages; ++i) {
			this._unseenPages[i] = true;
		}
		
		this._pageTokens = {};
		this._pageImages = {};
		this._gotPages = {};
	}
	
	serialize() {
		return JSON.stringify({
			bookTitle: this._bookTitle,
			usedQueries: _.keys(this._usedQueries),
			suggestedQueries: _.keys(this._suggestedQueries),
			unseenPages: _.keys(this._unseenPages),
			pageTokens: this._pageTokens,
			pageImages: this._pageImages,
			gotPages: _.keys(this._gotPages)
		});
	}
	
	deserialize(str) {
		const {
			bookTitle,
			usedQueries,
			suggestedQueries,
			unseenPages,
			pageTokens,
			pageImages,
			gotPages
		} = JSON.parse(str);
		
		this._bookTitle = bookTitle;
		
		this._usedQueries = {};
		_.each(usedQueries, (usedQuery) => {
			this._usedQueries[usedQuery] = true;
		});
		this._suggestedQueries = {};
		_.each(suggestedQueries , (suggestedQuery) => {
			this._suggestedQueries[suggestedQuery] = true;
		});
		
		this._unseenPages = {};
		_.each(unseenPages, (unseenPage) => {
			this._unseenPages[unseenPage] = true;
		});
		
		this._pageTokens = pageTokens;
		this._pageImages = pageImages;
		this._gotPages = {};
		_.each(gotPages, (gotPage) => {
			this._gotPages[gotPage] = true;
		});
	}
	
	getBookTitle() {
		return this._bookTitle;
	}

	addSuggestedQuery(query = '') {
		if (!this.usedQuery(query)) {
			this._suggestedQueries[query] = true;
		}
	}
	
	usedQuery(query = '') {
		return !!this._usedQueries[query];
	}
	
	onQuery(query = '') {
		delete this._suggestedQueries[query];
		this._usedQueries[query] = true;
	}
	
	onSuggestedQuery(query = '') {
		if (!this._usedQueries[query] && !this._suggestedQueries[query]) {
			this._suggestedQueries[query] = true;
		}
	}
	
	onPageTokens(pageTokens = {}) {
		_.each(pageTokens, (token, page) => {
			if (!this._gotPages[page] && !this._pageImages[page]) {
				delete this._unseenPages[page];
				this._pageTokens[page] = token;
			}
		});
	}
	
	onPageImages(pageImages) {
		_.each(pageImages, (imageUrl, page) => {
			if (!this._gotPages[page] && imageUrl) {
				delete this._unseenPages[page];
				delete this._pageTokens[page];
				this._pageImages[page] = imageUrl;
			}
		})
	}
	
	onPageDownload(page) {
		if (!this._gotPages[page]) {
			delete this._unseenPages[page];
			delete this._pageTokens[page];
			delete this._pageImages[page];
			this._gotPages[page] = true;
		}
	}
	
	getSuggestedQueries() {
		return _.keys(this._suggestedQueries);
	}
	
	getPageTokens() {
		return _.clone(this._pageTokens);
	}
	
	getPageImages() {
		return _.clone(this._pageImages);
	}
	
	toString() {
		return `unseen: ${_.size(this._unseenPages)}, tokens: ${_.size(this._pageTokens)}, img urls: ${_.size(this._pageImages)}, got: ${_.size(this._gotPages)}`;
	}
}
